Core/UsageSegmentationCalculator.swift

```diff
@@ -0,0 +1,372 @@
+    private func updateState(_ atb: Atb, pixelInfo: [String: Any]) {
+        previousAtb = atb
+
+        // # Trim history to 28d and add today
+        usageHistory = usageHistory.filter {
+            atb - $0 <= 28
+        } + [atb]
+
+        if pixelInfo[Params.countAsWAU] != nil {
+            let segments = pixelInfo[Params.segmentsToday] as? [String] ?? []
+            // # Filter out segments irrelevant to wau
+            previousWAUSegments = segments.filter {
+                return !$0.contains("_mau") && $0 != "first_month"
+            }
+        }
+
+        if let countAsMAUn = pixelInfo[Params.countAsMAUn] as? String {
+            let segments = pixelInfo[Params.segmentsToday] as? [String] ?? []
+            for n in 0 ..< 4 where countAsMAUn.safeCharAt(n) == "t" {
+                previousMAUSegments[n] = segments.filter {
+                    // # Filter out irrelevant to MAU
+                    return !$0.contains("_wau") && !$0.contains("_week")
+                    //  # Filter out irrelevant to _this_ MAU
+                    && ($0.contains("_mau_\(n)") || !$0.contains("_mau_"))
+                }
+            }
+        }
+    }
+
+    /// `get_segments`
+    private func getSegments(_ atb: Atb) -> [String] {
+        var segments: [String] = []
+
+        if countAsWAU(atb) {
+            if atb.week == installAtb.week {
+                segments.append("first_week")
+            } else if atb.week == installAtb.week + 1 {
+                segments.append("second_week")
+            } else if countsAsWAUAndActivePreviousWeek(atb) {
+                segments.append("current_user_wau")
+            } else {
+                segments.append("reactivated_wau")
+            }
+        }
+
+        if (previousAtb == nil || previousAtb == installAtb)
+            && atb.week < installAtb.week + 4 {
+            segments.append("first_month")
+        } else {
+            for n in 0 ..< 4 where countAsMAU(n, atb) {
+                if countsAsMAUAndActivePreviousMonth(n, atb) {
+                    segments.append("current_user_mau_\(n)")
+                } else {
+                    segments.append("reactivated_mau_\(n)")
+                }
+            }
+        }
+
+        if installAtb.isReturningUser
+            // # ATB cohorts get generalized after 28d
+            // # Hopefully this is handled elsewhere in the real code!
+            && atb.ageInDays <= installAtb.ageInDays + 28 {
+            segments.append("reinstaller")
+        }
+
+        if segmentRegular(atb) {
+            segments.append("regular")
+        }
+
+        if segmentIntermittent(atb) {
+            segments.append("intermittent")
+        }
+
+        return segments.sorted()
+    }
+
+    /// `count_as_wau`
+    private func countAsWAU(_ atb: Atb) -> Bool {
+        if atb == installAtb {
+            // # Install day - this code should not be running! Report nothing.
+            assertionFailure("See comment")
```

Comment:
Do we think it would be useful to log any of the `assertionFailure`s?

Reply:
Realistically, probably not.  If something is going wrong with this code we'd likely debug it locally anyway.  Also this will occur one time on one day so the chances of us catching logs with this in are pretty remote imo.